lib/extension/YOrmDatasetRelatedDataDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+declare(strict_types=1);
+
+namespace rexstan;
+
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\Generic\GenericObjectType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use rex_yform_manager_collection;
+use rex_yform_manager_dataset;
+use rex_yform_manager_query;
+use function count;
+use function in_array;
+
+final class YOrmDatasetRelatedDataDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
+{
+    public function getClass(): string
+    {
+        // @phpstan-ignore-next-line
+        return rex_yform_manager_dataset::class;
+    }
+
+    public function isMethodSupported(MethodReflection $methodReflection): bool
+    {
+        return in_array(
+            strtolower($methodReflection->getName()),
+            ['getrelateddataset', 'getrelatedcollection', 'getrelatedquery'],
+            true
+        );
+    }
+
+    public function getTypeFromMethodCall(
+        MethodReflection $methodReflection,
+        MethodCall $methodCall,
+        Scope $scope
+    ): ?Type {
+        $args = $methodCall->getArgs();
+        if (1 < count($args)) {
+            return null;
+        }
+
+        if (
+            !class_exists(rex_yform_manager_dataset::class)
+            || !class_exists(rex_yform_manager_collection::class)
+            || !class_exists(rex_yform_manager_query::class)
+        ) {
+            return null;
+        }
+
+        $key = $scope->getType($args[0]->value);
+        $constantStrings = $key->getConstantStrings();
+        if ($constantStrings === []) {
+            return null;
+        }
+
+        $datasetObject = $scope->getType($methodCall->var);
+        $classReflections = $datasetObject->getObjectClassReflections();
+        $method = strtolower($methodReflection->getName());
+
+        require_once __DIR__.'/../../test2.php';
```

Comment:
Ist das hier bewusst drin?

Reply:
das war der testcode beschrieben in https://github.com/FriendsOfREDAXO/rexstan/pull/408#issuecomment-1474857979

muss hier weg